### PR TITLE
Real estate Attempt 2

### PR DESCRIPTION
**Pull request by fgsahoward**
Players can purchase railroads, utilities and Properties. When a player lands on unowned real estate, they immediately purchase it and the price is deducted from their balance (you can allow the player's balance to go below 0 or you can cancel the purchase if they cannot afford it).

When a player lands on an owned real estate, they must pay rent equal to the rent amount to the owner (assuming they are not the owner). 

When a player lands on a mortgaged property, nothing happens.

A player has the option of mortgaging a property or paying off the mortgage of a property at the beginning of their turn or at the end of their turn.

Properties: If a player owns all of the properties in a color group, the rent doubles. 

Utilities: If only one utility is owned, then rent is equal to 4 times the value currently shown on the dice. If both utilities are owned (not necessarily by the same person), then rent is equal to 10 times the value currently shown on the dice.

Railroads: If a player owns one railroad, rent is $25. If a player owns two, rent is $50, 3 is $100 and 4 is $200.

### CODE REVIEW

File: Board.cs [hunk 1]
```diff
@@ -1,16 +1,93 @@
...
+using System.Linq;
+
+namespace Monopoly
 {
     public class Board : IBoard
     {
-        public IProperty[] Properties { get; set; }
-        public int PropertyCount
+        private Dictionary<string, IPropertyGroup> nameToGroupMapping;
+        private Dictionary<int, IPropertyGroup> mapIndexToGroupMapping;
```
Comment (FGS-BMarkham) [inline Board.cs]: These fields don't get (re-)assigned after the constructor runs, so you can mark them as `readonly`.

Comment (fgsahoward) [inline Board.cs]: Good call, to prevent messing up my commit history, I've fixed this in the last release

File: Game.cs [hunk 2]
```diff
@@ -28,18 +37,208 @@ private void PlayRound()
...
+            if (property.Mortgaged && player.Money > property.Price)
+            {
+                property.Mortgaged = false;
+                player.Money -= property.Price;
+                Console.WriteLine(
+                    "{0} unmortgaged {1}, net worth is now ${2}",
+                    player.Name,
+                    property.Name,
+                    player.Money);
+            }
```
Comment (FGS-BMarkham) [inline Game.cs]: While the `Dictionary<string, Action<IPlayer, IProperty>>` is really slick stuff, this class has bloated well past a Single Responsibility. Consider cases where multiple developers adding/changing minor game behavior and then trying to merge conflicts with this class. Also consider the need to specifically unit test the logic in methods like `MortgageProperty` & `UnMortgageProperty`, whose test coverage are currently only incidental.

Consider using the Command pattern, or the Observer pattern, (or a combination) to separate these disparate chunks of logic into single-purpose units.

File: Property.cs [hunk 3]
```diff
@@ -3,10 +3,26 @@
     public class Property : IProperty
     {
         public string Name { get; set; }
+        public IPlayer Owner { get; set; }
+        public int Price { get; }
+        public int Rent { get; }
+        public bool Mortgaged { get; set; }
+        public string Action { get; }
+        public int MapIndex { get; }
```
Comment (FGS-BMarkham) [inline Property.cs]: A `Property` with a bunch of C# _properties_ like this is an indication of a violation of "Tell, Don't Ask". Consider a design where these responsibilities are spread out. i.e. a "RealEstateAgent" class/abstraction could answer questions about whether or not a property is owned (and by whom), another class/abstraction could manage tracking mortgages, and a third class/abstraction could handle rent calculation - and they would all collaborate. This gives rise to centralized handling of those responsibilities, whereas right now, I need to track down all usages of all these properties to reconstruct how the system works.

File: PropertyGroupFactory.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using Newtonsoft.Json;
+
+namespace Monopoly
+{
+    public class PropertyGroupFactory
+    {
+        public static PropertyGroup[] GenerateGroups(string allData)
+        {
+            return JsonConvert.DeserializeObject<PropertyGroup[]>(allData, new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.Auto });
```
Comment (FGS-BMarkham) [inline PropertyGroupFactory.cs]: Be still my heart. Json! :+1:

File: PropertyGroupTests.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    public class PropertyGroupTests
+    {
+        private IPropertyGroup group;
+        private IProperty[] properties;
+        private Player[] players;
+
+        [SetUp]
+        public void Setup()
+        {
+            RandomGeneratorMoc generator = new RandomGeneratorMoc();
```
Comment (FGS-BMarkham) [inline PropertyGroupTests.cs]: Misspelled "mock". :-)

File: json/propertyGroups.json [hunk 6]
```diff
@@ -0,0 +1,364 @@
+[
+    {
+        "Properties": [
+            {
+                "$type": "Monopoly.Property, ConsoleApplication",
```
Comment (FGS-BMarkham) [inline json/propertyGroups.json]: You probably shouldn't need this if you aren't (de)serializing these with a polymorphic context.

Comment (fgsahoward) [inline json/propertyGroups.json]: This is the only way that Newtonsoft.JSON can deserialize the inner-most dictionaries into "Property" objects. When i remove it, the PropertyFactory breaks trying to figure out what type to cast those to. Trust me, I really didn't want that there, but it's the only way I could make it work

File: Property.cs (hunk 3 above)
Comment (FGS-BMarkham) [inline Property.cs]: Note that how the properties are (de)serialized quickly becomes orthogonal to how they are handled at runtime, in which case this existing implementation serves well as the _persistence_ model for properties.

File: json/propertyGroups.json (hunk 6 above)
Comment (FGS-BMarkham) [inline json/propertyGroups.json]: While I said it in a different context, it applies here, too:
> Note that how the properties are (de)serialized quickly becomes orthogonal to how they are > handled at runtime.

If you are attached to the serialization route, consider using dedicated models to represent the on-disk format, which will avoid this problem and also make the rest of the app less coupled to the file format.

**Comment by FGS-BMarkham** [comment]
Please also work on keeping your PRs smaller and cleaner. In general, 19 commits in a PR for one of our products like 'PUMP' is a surefire way to make your coworkers procrastinate reviewing it, which is a recipe for frustration.

The commit history should "tell a story" about how a business or technical problem was solved, but shouldn't contain any kind of thrashing. (A commit that fixes a problem introduced by a previous in the same PR should never actually make it to PR.) Merge commits should never be visible in PRs, as rebasing avoids the need. Also consider keeping commit messages in a more purposeful format, such as subject-body style, to make it easier for reviewers to either skim or deeply investigate the intent of individual commits.

**Pull request closed without merging**